src/assembler/parser/index.ts

```diff
@@ -0,0 +1,91 @@
+import { AreaType, IArea, ICodeFile, IInstruction } from 'assembler/ast'
+import { ParseError } from './error'
+import { ITextMatch, ITextParseRule, parseText } from './text'
+
+/**
+ * Parses a given code string and return a parsed code file (AST representation).
+ *
+ * @param code the assembler code (text representation)
+ * @returns the parsed code file (AST representation)
+ */
+export function parse(code: string): ICodeFile {
+  const ast: ICodeFile = {
+    symbols: {},
+    areas: []
+  }
+  let area: IArea | null = null
+  let label: string | null = null
+  const rules: ITextParseRule[] = [
+    {
+      name: 'Whitespace',
+      pattern: /\s+/
+    },
+    {
+      name: 'Comment',
+      pattern: /;[^\n]*/
+    },
+    {
+      name: 'LiteralSymbolDeclaration',
+      pattern: /([a-z\|\._]+) +EQU +([0-9a-z#]+)/,
+      onMatch(match: ITextMatch) {
+        ast.symbols[match.captures[0]] = match.captures[1]
+        label = null
+      }
+    },
+    {
+      name: 'AreaDeclaration',
+      pattern: /AREA +([a-z\|\._]+) *, *(DATA|CODE) *, *(READ(WRITE|ONLY))/,
+      onMatch(match: ITextMatch) {
+        area = {
+          name: match.captures[0],
+          type:
+            match.captures[1].toUpperCase() === 'CODE'
+              ? AreaType.Code
+              : AreaType.Data,
+          isReadOnly: match.captures[2].toUpperCase() === 'READONLY',
+          instructions: []
+        }
+        ast.areas.push(area)
+        label = null
+      }
+    },
+    {
+      name: 'Label',
+      pattern:
+        /([a-z0-9]+)(?=\s+([a-z]+) +([0-9a-z#\[\]]+(, *[0-9a-z#\[\]]+)))/,
```

Comment:
Can a label start with a digit?

Reply:
underscore is valid according to keil uvision but digits are not. See:
![grafik](https://user-images.githubusercontent.com/56474048/164383512-02cc960c-5242-4567-9d45-4556fb05bb92.png)

Reply:
I have found this documentation from ARM Keil which states:
```
symbol must begin in the first column. It cannot contain any white space character such as a space or a tab unless it is enclosed by bars (|).
Labels are symbolic representations of addresses. You can use labels to mark specific addresses that you want to refer to from other parts of the code. Numeric local labels are a subclass of labels that begin with a number in the range 0-99. Unlike other labels, a numeric local label can be defined many times. This makes them useful when generating labels with a macro.
```
See https://www.keil.com/support/man/docs/ARMASM/armasm_dom1359731141352.htm

There is also a description about the label that is used for an area declaration:
```
You can choose any name for your sections. However, names starting with a non-alphabetic character must be enclosed in bars or a missing section name error is generated. For example, |1_DataArea|.
```

See: https://www.keil.com/support/man/docs/ARMASM/ARMASM_dom1361290002714.htm

So according to this, it should be possible if you enclose the label in bars ||.

Reply:
But maybe this is for another compiler because I just tried it and it does not really work in our Keil uVision so I guess only alphanumeric characters are allowed then...

Reply:
I've updated the parser with the latest regex.